x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/DataTypeConversion.java

```diff
@@ -73,7 +74,28 @@ public static DataType commonType(DataType left, DataType right) {
                 return left;
             }
         }
+
         // interval and dates
+        if (left == DATETIME) {
```

Comment:
I think this can be improved an currently allows more cases then it should.

With a date and interval, the date should always be on the left.
That's because date - interval works, but interval - date does not. It's something that I see most other DBs use as well.

Also please keep the current style which is simpler to read - that is only one type check per if.
`if (left == DATE) { if (right == DATETIME) ... if (DataType.isInterval(...)}`

Reply:
Please check: https://github.com/elastic/elasticsearch/pull/37693/commits/b4109031e770ac52153184b32baf921c112e8acd#diff-51f788d55d8ade70340fdf2ca8af1f20R86

Reply:
Tested and it's caught as an error here: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessor.java#L79

Reply:
That's not enough though - the case should be caught in the type resolution and reported accordingly. Currently the type resolution passes and an exception is thrown instead (which typically would occur if the plan passes but somehow the underlying data is not correct).

Reply:
I don't get this, if we want to support `INTERVAL + date` we need that.
The commonTypes() is not the place to check this imho, this should be checked specifically for the Subtrack operation for example.

Reply:
I'm not suggesting to fix this in `commonType` - instead that the type resolution of `Sub` should add an extra check.
Before this PR `commonType` did not allow the interval to be on the left and the date on the right and `Sub` was the reason; if `commonType` is changed then Sub needs to be updated.

Reply:
okidoki, I misundertood you, sorry, adding the check and a test.